tests/push/test_push.py

```diff
@@ -129,15 +129,18 @@ def test_typical_push(
     )
 
     # It should have invoked hook(s).
-    assert len(hookspy) == 11
+    assert len(hookspy) == 24
     (hook_name, hook_kwargs) = hookspy[0]
     assert hook_name == "task_start"
     (hook_name, hook_kwargs) = hookspy[1]
     assert hook_name == "pulp_repository_pre_publish"
     (hook_name, hook_kwargs) = hookspy[2]
     assert hook_name == "pulp_repository_published"
-    (hook_name, hook_kwargs) = hookspy[-2]
+    # after pulp_repository_published there's 13 calls of pulp_item_finished
```

Comment:
There should be at least one test which verifies the actual data passed into the hook, otherwise it can change without warning and break anything connected to the hook.

Reply:
...and by actual data I mean something more precise than mock.ANY.

Reply:
Hmm, but that's kinda changing unit test to integration test isn't it? When hook is called with attrs.asdict(item.pulp_unit). When you add or remove any attributes in pulp unit tests will fail.

Reply:
I'm not interested in trying to label each test as a certain type (unit/integration/system). I'm more interested in which bugs the test suite can detect.

In this case if the arguments weren't tested at all or were tested as mock.ANY, then even critical and obvious bugs such as calling the hook with swapped arguments would still pass all tests.

I think the way you've updated it now, where the arguments are tested but just lightly by only looking at some of the fields or types, is a good middle ground.  As it's quite unlikely there could be a bug where the hook is invoked with an object which is simultaneously a PushItem, but the wrong PushItem.

So I think the approach right now is decent, but it will need some changes to align with the other suggestions about yielding actual PulpUnit instances rather than dicts.